Feature/animations

aralioglu opened the pull request:
Added animation script with animation sprite cuts. Movement method adjust to ensure compatibility with animations.

aralioglu committed:
animations.py: Added initial animations containing movement, melee attack, ranged attack and spell cast

aralioglu committed:
Modified game_window, player and movement scripts to ensure compatibility with animations

AlpSari commented on omg/mechanics/animations.py:
```diff
@@ -0,0 +1,275 @@
+    
+    def update_animation(self, delta_time):
+
+        # Figure out if we need to flip face left or right
+        if self.player_change_y < 0 and self.character_face_direction != DOWN_FACING:
+            self.character_face_direction = DOWN_FACING
+            self.idle_textures = self.idle_textures_down
+            self.walk_textures = self.walk_textures_down
+            self.slash_textures = self.slash_textures_down
+            self.spellcast_textures = self.spellcast_textures_down
+            self.thrust_textures = self.thrust_textures_down
+            self.shoot_textures = self.shoot_textures_down
+        if self.player_change_y > 0 and self.character_face_direction != UP_FACING:
+            self.character_face_direction = UP_FACING
+            self.idle_textures = self.idle_textures_up
+            self.walk_textures = self.walk_textures_up
+            self.slash_textures = self.slash_textures_up
+            self.spellcast_textures = self.spellcast_textures_up
+            self.thrust_textures = self.thrust_textures_up
+            self.shoot_textures = self.shoot_textures_up
+        if self.player_change_x < 0 and self.character_face_direction != LEFT_FACING:
+            self.character_face_direction = LEFT_FACING
+            self.idle_textures = self.idle_textures_left
+            self.walk_textures = self.walk_textures_left
+            self.slash_textures = self.slash_textures_left
+            self.spellcast_textures = self.spellcast_textures_left
+            self.thrust_textures = self.thrust_textures_left
+            self.shoot_textures = self.shoot_textures_left
+        if self.player_change_x > 0 and self.character_face_direction != RIGHT_FACING:
+            self.character_face_direction = RIGHT_FACING
+            self.idle_textures = self.idle_textures_right
+            self.walk_textures = self.walk_textures_right
+            self.slash_textures = self.slash_textures_right
+            self.spellcast_textures = self.spellcast_textures_right
+            self.thrust_textures = self.thrust_textures_right
+            self.shoot_textures = self.shoot_textures_right
```
if we define that a `Texture` class mentioned in the constructor, we can also simplify the logic here. Let us discuss after we fixed the previous part.

AlpSari commented on omg/mechanics/movement.py:
```diff
@@ -21,6 +22,10 @@ def __init__(self, forward, backward, left, right):
         self.pressed_right: bool = False
 
+        # Track animations
+        self.move_direction = (0,0)
```
Indeed, the `move_direction` should be part of this class and it should be propagated to the `Animations` class.

AlpSari commented on omg/mechanics/movement.py:
```diff
@@ -42,41 +47,59 @@ def on_key_press(self, key, modifiers):
         if key == self.key_forward:
             self.pressed_forward = False
+            if not self.pressed_backward:
+                self.move_direction = (self.move_direction[0], 0)
```
Tuples are immutable objects in Python. Therefore, you need to set entire tuple whenever you need to change the tuple. Then, my question is: can we use another data structure such as `List` here?

AlpSari commented on omg/mechanics/movement.py:
```diff
@@ -126,18 +149,21 @@ def calculate_player_state(
-        angle_rad = self.face_mouse(mouse_x, mouse_y, center_x, center_y)
-        angle = math.degrees(angle_rad)
-        return vx_wrt_ground, vy_wrt_ground, angle
+        angle = 0   #Angle is always 0 to enable proper animations
```
The problem here is: this `angle` dictates both character_angle` and the `shooting_angle`. 

You wanted to only constrain `char_angle`. That means we need two `angle` variables one defining `character_angle` and the other one `shooting_angle`. The latter can be calculated with the mouse position.

AlpSari reviewed (changes requested):
Great work ! There are some implementation considerations which can make the code even shine more.

aralioglu committed:
Adjusted sprite names for future use and scalability
Updated texture generation for general purpose use
Fixed a bug in the movement script

AlpSari committed:
mechanics: Added animation class to represent a single action in all directions

AlpSari committed:
mechanics: refactor animations module to use newly introduced animation class, remove dependency to textures

AlpSari committed:
mechanics: refactor textures (not used anymore)

AlpSari committed:
mechanics: remove textures.py, which is deprecated

AlpSari committed:
mechanics: remove deprecated main() in animation

AlpSari committed:
mechanics: remove deprecated imports

aralioglu committed:
Updated animations.py on_key_press function for better scalability

aralioglu committed:
Finally done!

aralioglu committed:
Cleaned unneccessary comments

aralioglu committed:
remove deprecated files

aralioglu committed:
rename animation image folder

aralioglu committed:
mechanics: update folder name

aralioglu committed:
Fix linting errors

aralioglu committed:
Fixed linting

aralioglu committed:
added changelog fragment

aralioglu committed:
mechanics: fix docstring linting errors

AlpSari committed:
Merge branch 'main' into feature/animations

AlpSari committed:
core: fix linting after merge conflict resolve

AlpSari committed:
Fix player constructor

AlpSari committed:
entities: remove comments

AlpSari committed:
animations are loaded from the player argument
